### PR TITLE
Add `with_halos` option to `Average` diagnostic

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR adds a `with_halos` option to the `Average` diagnostic. By default `with_halos = false`. Previously `with_halos` was always `true` which was awkward as most use cases did not warrant halos in the output.

PR test `Average` with halos and add result size tests.

### CODE REVIEW

File: src/Diagnostics/average.jl [hunk 1]
```diff
@@ -81,5 +84,6 @@ end
 
 function (avg::Average)(model)
     run_diagnostic(model, avg)
-    return avg.return_type(avg.result)
+    N, H = size(model.grid), halo_size(model.grid)
+    return avg.return_type(avg.result)[(avg.with_halos ? Colon() : d in avg.dims ? Colon() : (1+H[d]:N[d]+H[d]) for d in 1:3)...]
```
Comment (ali-ramadhan) [inline src/Diagnostics/average.jl:88]: This is probably overkill and could be much more readable if converted to an `if` statement with good comments...

**Comment by ali-ramadhan** [comment]
I agree with your four points and that a default of `with_halos = true` makes sense in anticipation of a more powerful fields abstraction. I'm happy to change the default (will commit once tests pass), I just wanted the option of including or excluding halos for different use cases.

**Comment by ali-ramadhan** [comment]
Hmmm, this has been a weird PR. Do not merge yet.

Was generating some training data when I noticed that horizontal averages were coming back as nonsense (well, horizontal average at iteration 0 made sense but not subsequent ones, e.g.:

```
>>> ds.T
<xarray.DataArray 'T' (time: 4, zC: 256)>
array([[1.900195e+001, 1.900586e+001, 1.900977e+001, ..., 1.999023e+001,
        1.999397e+001, 1.988585e+001],
       [0.000000e+000, 5.582942e-322, 2.699495e-315, ..., 2.212087e+214,
        2.602799e+243, 2.521740e+180],
       [2.069993e+161, 3.671430e+228, 2.696493e+161, ..., 5.451647e+247,
        1.335554e+185, 5.821384e+233],
       [2.908794e+257, 3.313691e+150, 4.983208e+151, ..., 2.023242e-076,
        1.142199e+243, 9.832549e-072]])
Coordinates:
  * zC       (zC) float64 -99.8 -99.41 -99.02 -98.63 ... -0.9766 -0.5859 -0.1953
  * time     (time) float64 0.0 603.0 1.206e+03 1.803e+03
Attributes:
    units:     °C
    longname:  Horizontally averaged temperature
```

It's fine if I go back to the master branch.

In debugging this issue I ran into https://github.com/JuliaGPU/CUDA.jl/issues/370 but I don't think it's related to this issue.

**Comment by glwagner** [comment]
Sounds good, maybe re-request a review when its ready?

**Comment by ali-ramadhan** [comment]
PR is ready to be reviewed/merged now I think.

Pretty sure the main issue turned out to be a problem with saving the `1×1×Nz` result arrays from calling `(::Average)(model)` to NetCDF variables with dimensions `("xC", "time")`, so I should have been dropping singleton dimensions before writing to NetCDF.

Once I fixed this by calling `dropdims` before writing `Average` results to NetCDF, the problem disappeared (see below).

I wasn't able to reproduce the issue in isolation with NCDatasets.jl at the REPL and not sure why it only became an issue in this PR...

Instead of calling `dropdims` before writing data to NetCDF, an alternative could be to provide a `reshape` option to the `Average` diagnostic which could be seen as introducing bloat into the `Average` diagnostic but it may have other uses in the future.

---

```
<xarray.DataArray 'T' (time: 4, zC: 256)>
array([[19.001953, 19.005859, 19.009766, ..., 19.990234, 19.994141, 19.998047],
       [19.001953, 19.005859, 19.009766, ..., 19.990234, 19.99413 , 19.960477],
       [19.001953, 19.005859, 19.009766, ..., 19.990234, 19.994097, 19.922942],
       [19.001953, 19.005859, 19.009766, ..., 19.990235, 19.993974, 19.885847]])
Coordinates:
  * zC       (zC) float64 -99.8 -99.41 -99.02 -98.63 ... -0.9766 -0.5859 -0.1953
  * time     (time) float64 0.0 603.0 1.206e+03 1.803e+03
Attributes:
    units:     °C
    longname:  Horizontally averaged temperature
```

**Pull request merged**